lib/Stan_math_code_gen.ml

```diff
@@ -116,61 +136,115 @@ let pp_arg ppf ((adtype, name, st), idx) =
   in
   pf ppf "const %a& %s" (pp_unsizedtype adstr) st name
 
-let pp_template_decls ppf =
-  pf ppf "@[<hov>template <%a>@]@ " (list ~sep:comma (fmt "typename %s"))
-
 let with_idx lst = List.(zip_exn lst (range 0 (length lst)))
 
 let%expect_test "with idx" =
   print_s [%sexp (with_idx (List.range 10 15) : (int * int) list)] ;
   [%expect {| ((10 0) (11 1) (12 2) (13 3) (14 4)) |}]
 
-let pp_error_wrapper ppf (pp_err, err_arg, pp_contents, contents_arg) =
-  string ppf "try " ;
-  pp_block ppf (pp_contents, contents_arg) ;
-  string ppf " catch const std::exception& e) " ;
-  pp_block ppf (pp_err, err_arg)
+let pp_decl ppf (vident, st) =
+  pf ppf "%a %s;" pp_prim_stantype (Ast.remove_size st) vident
+
+let with_no_loc stmt = {stmt; sloc= ""}
 
 let pp_located_msg ppf msg =
   pf ppf
     {|stan::lang::rethrow_located(
-    std::runtime_error(std::string(%s) + e.what(), current_statement__);
-// Next line prevents compiler griping about no return
-throw std::runtime_error("*** IF YOU SEE THIS, PLEASE REPORT A BUG ***");
-|}
+          std::runtime_error(std::string(%s) + e.what(), current_statement__));
+      // Next line prevents compiler griping about no return
+      throw std::runtime_error("*** IF YOU SEE THIS, PLEASE REPORT A BUG ***"); |}
   @@ Option.value ~default:"e" msg
 
-let pp_located_error ppf (pp_contents, contents_arg, msg) =
-  pp_error_wrapper ppf (pp_located_msg, msg, pp_contents, contents_arg)
+let maybe_templated_arg_types (args : formal_params) =
+  let is_autodiff (adtype, _, _) =
+    match adtype with Ast.AutoDiffable -> true | _ -> false
+  in
+  List.filter ~f:is_autodiff args |> List.mapi ~f:(fun i _ -> sprintf "T%d__" i)
+
+let%expect_test "arg types templated correctly" =
+  [(Ast.AutoDiffable, "xreal", Ast.UReal); (Ast.DataOnly, "yint", Ast.UInt)]
+  |> maybe_templated_arg_types |> String.concat ~sep:"," |> print_endline ;
+  [%expect {| T0__ |}]
+
+(** Pretty-prints a function's return-type, taking into account templated argument
+    promotion.*)
+let pp_returntype ppf arg_types rt =
+  let scalar =
+    strf "typename boost::math::tools::promote_args<@[<-35>%a@]>::type"
+      (list ~sep:comma string)
+      (maybe_templated_arg_types arg_types)
+  in
+  pf ppf "%a@ " (option ~none:(const string "void") (pp_unsizedtype scalar)) rt
 
-let pp_decl ppf (vident, st) =
-  pf ppf "%a %s;" pp_prim_stantype (Ast.remove_size st) vident
+let pp_location = fmt "current_statement__ = %S;@;"
 
-let with_no_loc stmt = {stmt; sloc= ""}
+(** [pp_located_error ppf (body_block, err_msg)] surrounds [body_block]
+    with a C++ try-catch that will rethrow the error with the proper source location
+    from the [body_block] (required to be a [stmt_loc Block] variant).*)
```

Comment:
What is a `stmt_loc Block` variant? What happens if you feed it something else?

Reply:
Hm, this has been a thorny thing in a few places - trying to know when we will be getting a Block (and thus printing braces) vs. not getting one. I put in a comment to try to refactor it generally (also mixes with your coalesce comments).